File: problem3.js
// ## Problem3 
// (https://leetcode.com/problems/container-with-most-water/)

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

const maxArea = function(height) {
    if(!height || height.length === 0) return 0;
    let j = height.length - 1, i =0, temp = 0;
    let max = (j - i) * Math.min(height[i], height[j]);
    while(i <= j) {
        if(height[i] < height[j]) i++;
        else j--;
        temp = (j - i) * Math.min(height[i], height[j]);
        if(temp > max) max = temp;
    }
    return max;
};

File: problem_3.cpp
/ Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
    int mini(int a , int b) {
        if (a<b) return a;
        else return b;
    }
    
    int maxi(int a, int b) {
        if (a>b) return a;
        else return b;
    }
public:
    int maxArea(vector<int>& height) {
        //max water = width * height ( ie, min of two heights)
        int l = 0;
        int h = height.size() - 1;
        int max_cap = 0;
        while (l < h) {
            int width = (h - l);
            int heightt = mini(height[l], height[h]); //todo
            int cap = width * heightt;
            max_cap = maxi(max_cap, cap);
            if (height[l] <= height[h]) l++;
            else h--;
        }
        
        return max_cap;
    }
};
